src/test/ui/impl-trait/dyn-trait-return-should-be-impl-trait.stderr

```diff
@@ -0,0 +1,186 @@
+  --> $DIR/dyn-trait-return-should-be-impl-trait.rs:16:25
+   |
+LL | fn ban() -> dyn Trait { Struct }
+   |             ---------   ^^^^^^ expected trait `Trait`, found struct `Struct`
+   |             |
+   |             expected `(dyn Trait + 'static)` because of return type
+   |
+   = note: expected trait object `(dyn Trait + 'static)`
+                    found struct `Struct`
+
+error[E0746]: return type cannot have a bare trait because it must be `Sized`
+  --> $DIR/dyn-trait-return-should-be-impl-trait.rs:16:13
+   |
+LL | fn ban() -> dyn Trait { Struct }
+   |             ^^^^^^^^^ doesn't have a size known at compile-time
+   |
+   = note: for information on `impl Trait`, see <https://doc.rust-lang.org/book/ch10-02-traits.html#returning-types-that-implement-traits>
+help: you can use the `impl Trait` feature in the return type because all the return paths are of type `Struct`, which implements `dyn Trait`
+   |
+LL | fn ban() -> impl Trait { Struct }
+   |             ^^^^^^^^^^
+
+error[E0277]: the size for values of type `(dyn Trait + 'static)` cannot be known at compilation time
+  --> $DIR/dyn-trait-return-should-be-impl-trait.rs:19:13
+   |
+LL | fn bak() -> dyn Trait { unimplemented!() }
+   |             ^^^^^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `std::marker::Sized` is not implemented for `(dyn Trait + 'static)`
+   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
+   = note: the return type of a function must have a statically known size
+
+error[E0308]: mismatched types
+  --> $DIR/dyn-trait-return-should-be-impl-trait.rs:23:16
+   |
+LL | fn bal() -> dyn Trait {
+   |             --------- expected `(dyn Trait + 'static)` because of return type
+LL |     if true {
+LL |         return Struct;
+   |                ^^^^^^ expected trait `Trait`, found struct `Struct`
+   |
+   = note: expected trait object `(dyn Trait + 'static)`
+                    found struct `Struct`
+
+error[E0746]: return type cannot have a bare trait because it must be `Sized`
+  --> $DIR/dyn-trait-return-should-be-impl-trait.rs:21:13
+   |
+LL | fn bal() -> dyn Trait {
+   |             ^^^^^^^^^ doesn't have a size known at compile-time
+   |
+help: if all the returned values were of the same type you could use `impl Trait` as the return type
+  --> $DIR/dyn-trait-return-should-be-impl-trait.rs:25:5
+   |
+LL |         return Struct;
+   |                ^^^^^^
+LL |     }
+LL |     42
+   |     ^^
+   = help: alternatively, you can always create a new `enum` with a variant for each returned type
+   = note: for information on `impl Trait`, see <https://doc.rust-lang.org/book/ch10-02-traits.html#returning-types-that-implement-traits>
+   = note: for information on trait objects, see <https://doc.rust-lang.org/book/ch17-02-trait-objects.html#using-trait-objects-that-allow-for-values-of-different-types>
```

Comment:
Note should be after the `help:` below.

Reply:
Suggestions always come last today, they don't keep correct track of their position. If prefer to do that in a separate PR.

Reply:
I would prefer having each (suggestion, explanation, links) pair be located together as it reduces mental context switching between the different suggestions. It also makes sense in terms of the code.